### ISSUE TITLE
Filter request/response bodies with JQ

### ISSUE THREAD

**Issue by pimterry** [httptoolkit/httptoolkit#115]
Make it possible to filter the response body, so you can quickly find content e.g. using `.[1000].name` to jump to the `name` field of the 1001st element in the response.

---

_Does this affect you too? Click the +:smiley: icon in the top right and add a :+1: on issues to help decide where HTTP Toolkit goes next, or [go take a look & vote on the most popular ideas so far](https://github.com/httptoolkit/feedback/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc)._

**Comment by daluu** [comment]
I don't know if mentioning `jq` is the ideal request here, that's a CLI tool and not sure it integrates well with httptoolkit. However, JSONPath might be something that would work (JSON version of XPath, for querying JSON). I haven't checked but hopefully jq & JSON path share the same syntax, as it is good to follow a standard query protocol than trying to support a specify query tool.

**Comment by pimterry** [comment]
Good point @daluu! Do you know of any stats about popularity of `jq` and tools using `jpath`? I do think `jq` is super popular, but you're right that it's not a general standard as such, and I'm open to alternatives. Overall I'd like to use the most widely recognized syntax I think.

**Comment by daluu** [comment]
Oh, I didn't notice this post was from the project's author, lol.

I'd be fine with jq, or JSONPath, but I haven't really delved into things to know which would work better. I was looking at this from the standpoint of:

* a user and what I may be familiar with, with regards to other tools. I think I recall JMeter supports JSONPath, not sure about jq. Gatling, also supports JSONPath, and more on that shortly.

*the httptoolkit developer and having to support the filtering query syntax, to which certain questions arise like: do I use a standard query syntax? Do I use existing library to give me the filter query ability? Do I write my own filter query parser for JSON?

Your questions made me think a bit and I did some online search and found some interesting info:

* A post from Gatling that talks about jq, JSONPath, and another new option: https://gatling.io/2019/07/31/introducing-jmespath-support/

* Another post. comparing the 3 JSON query options: https://levelup.gitconnected.com/json-queries-give-your-users-jmespath-power-ef8ab0d38553

* More mentions of JMESPath https://blog.scottlowe.org/2018/06/28/more-handy-cli-tools-json/

I leave it up to you to decide what seems most appropriate for HTTPToolkit in terms of being a user and as the developer/maintainer.

**Comment by erhhung** [comment]
Would it be possible, and desirable, to configure generic, externally installed CLI tools that could allow the user to filter any kind of response through one of the configured tools, and not limited to builtin query syntaxes, like JSONPath for JSON, and XPath for XML?

For many cloud developers these days, people use `jq`, JMESPath (AWS), JSONPath (K8s), etc. to query JSON, and if I deal with legacy XML output, I like to use `oq` to filter using `jq` syntax which I'm much more familiar with.

If we can configure a few command-line tools with a query placeholder argument and which response content types it supports, and that can accept stdin and return stdout, then on any particular response, let the user click a button next to "Filter response through: \<dropdown\>", which shows a dialog or pane where the user can see the original response and possibly enter a query that will be fed to the selected external process.

There's certainly more details I've overlooked, but a generic interface to filter any response, maybe even an image through `exiftool`, would be great.

**Comment by daluu** [comment]
Maybe this is something of a feature request and poll the (potential) user community to see how they would like to see things implemented?

**Comment by pimterry** [comment]
Interesting!

I'm cautious to shell out to external tools arbitrarily, mostly for performance reasons, and to avoid complex configuration scenarios. While it's not going to be _super_ slow I suspect launching processes and passing data around between them will add a little latency that makes doing any kind of filtering-as-you-type approaches painful, and it makes it hard to do things like syntax highlighting or autocomplete for a syntax inside the app extremely difficult. It's not impossible, but it limits the options, and I don't think it's necessary.

That's not to say we can't support more than one syntax regardless though, and all of these seem to be available as pure-JS libraries, so we can implement any of them natively, in process. I suspect the right answer is to start with the most popular one (either from public existing-usage stats if anybody can find them, looking at other tools in the ecosystem, or polling the community) and then to look into #56 to support further syntaxes via plugins, to start with (later maybe integrating any super popular plugins directly).
